Added ios & android technique and tool for re-flutter (by @appknox) (#2600)
OWASP/owasp-mastg: techniques/android/MASTG-TECH-0109.md

--- 

> title: Intercepting Flutter HTTP Traffic

Reviewer: You're correct! Thanks

platform: android 
---

Flutter is an open-source UI software development kit (SDK) created by Google. It is used for building natively compiled applications for mobile, web, and desktop from a single codebase. Flutter uses Dart, which is not proxy-aware and uses its own certificate store. The application doesn't take proxy configuration from the system and send the data directly to the server. Due to this, it is not possible to intercept the request using the BurpSuite or any MITM tools.

> **How does re-flutter method differs from other techniques ?**

Reviewer: @sk3l10x1ng thanks for all the updates, I've fixed a couple of typos for now.

Also thanks for checking those blog posts. I think the only think that remains open is the consideration of the 3rd alternative technique. If I'm not misunderstanding anything they are 2 alternative approaches, 
- (re-flutter) requires patching 
- (nviso script) doesn't require patching, it uses Frida

Using these apps:
- https://github.com/NVISOsecurity/disable-flutter-tls-verification/blob/main/test_app/pinning.apk
- https://github.com/NVISOsecurity/disable-flutter-tls-verification/blob/main/test_app/pinning.ipa

Could you please validate that your technique works with them and also using this alternative technique via Frida, it seems easy to run using codeshare (as indicated [here](https://github.com/NVISOsecurity/disable-flutter-tls-verification)):

```
frida -U --codeshare TheDauntless/disable-flutter-tls-v1 -f YOUR_BINARY
```

If all of this works we can confirm that both techniques are equivalent and could be summarized as:

## re-flutter
1. setup interception proxy
2. patch app using re-flutter
3. inspect intercepted traffic

## nviso
1. setup interception proxy
2. run frida script
3. inspect intercepted traffic

Reviewer: @TheDauntless Would you mind chiming in to confirm my hypothesis (or not)?

> There are alternative methods for intercepting traffic, such as [sending traffic to the proxy through ProxyDroid/iptables](https://blog.nviso.eu/2019/08/13/intercepting-traffic-from-android-flutter-applications/). However, these techniques require some configuration. By employing the re-flutter command-line tool, the application can be patched effortlessly without the need for any setup.

Reviewer 2: ```suggestion
In order to intercept Flutter HTTPS traffic we need to deal with two problems:
* Make sure the traffic is sent to the proxy
* Disable the TLS verification of any HTTPS connection

There are generally two approaches to this: reFlutter and Frida. 

* reFlutter: This tool creates a modified version of libFlutter.so which is then repackaged into the APK. It configures the internal libraries to use a specified proxy and disable the TLS verification
* Frida: The [disable-flutter-tls.js script](https://github.com/NVISOsecurity/disable-flutter-tls-verification) can dynamically remove the TLS verification without the need for repackaging. As it doesn't modify the proxy configuration, additional steps are needed (e.g. ProxyDroid, DNS, iptables, ...)
```
The re-flutter app also has downsides, so this is pretty one-sided:
* Only works on known/published hashes, otherwise you have to rebuild the engine yourself
* Repackages the app. This brings complications (e.g. iOS or any app that detects repackaging)

I've used reFlutter a few times (mostly for the object-dump) and it's great when it works, but not straightforward if it doesn't.

So I modified this section to give a more generic introduction of what needs to be done, and then the rest can explain both reFlutter and Frida+(eg)ProxyDroid

We should also use reFlutter, as the tool calls itself, and not re-flutter.

## Intercepting Traffic using re-flutter

1. Patch the app to enable traffic interception.

> Run the command to patch the app and select the option **Traffic monitoring and interception** and then the IP of the machine which the interception proxy is running.

Reviewer 2: ```suggestion
Run the command to patch the app and select the option **Traffic monitoring and interception** and then enter the IP of the machine on which the interception proxy is running.
```

```
$ reflutter demo.apk

Choose an option:

    Traffic monitoring and interception
    Display absolute code offset for functions

[1/2]? 1

Example: (192.168.1.154) etc.
Please enter your BurpSuite IP: 192.168.29.216
```

This will create a **release.RE.apk** file in the output folder.

2. Sign the patched **release.RE.apk** file (e.g. using the [uber-apk-signer](https://github.com/patrickfav/uber-apk-signer)).

```
   $ java -jar uber-apk-signer.jar -a release.RE.apk --out demo-signed
```

This will create a **release.RE-aligned-debugSigned.apk** file in the output folder.

3. Install the signed patched app on the mobile device.

> 4. Configure the interception proxy.For example, in Burp-suite:

Reviewer 2: ```suggestion
4. Configure the interception proxy. For example, in Burp:
```
We use 'Burp' as the consistent name for PortSwigger's Burp Suite throughout the MASTG.

  - Under Proxy -> Proxy settings -> Add new Proxy setting.

>   - Bind listening Port to 8083.

Reviewer 2: ```suggestion
  - Bind listening Port to `8083`.
```

>   - Select Bind to address to All interfaces.

Reviewer 2: ```suggestion
  - Select `Bind to address` to `All interfaces`.
```

```plaintext
$ reflutter demo.apk

Choose an option:

    Traffic monitoring and interception
    Display absolute code offset for functions

[1/2]? 1

Example: (192.168.1.154) etc.
Please enter your BurpSuite IP: 192.168.29.216
```

This will create a **release.RE.apk** file in the output folder.

2. Sign the patched **release.RE.apk** file (e.g. using the [uber-apk-signer](https://github.com/patrickfav/uber-apk-signer)).

```plaintext
java -jar uber-apk-signer.jar -a release.RE.apk --out demo-signed
```

This will create a **release.RE-aligned-debugSigned.apk** file in the output folder.

3. Install the signed patched app on the mobile device.

4. Configure the interception proxy. For example, in Burp:

- Under Proxy -> Proxy settings -> Add new Proxy setting.
- Bind listening Port to `8083`.
- Select `Bind to address` to `All interfaces`.
- Request Handling -> support for invisible proxying.

5. Open the app and start intercepting traffic.

## Intercepting Traffic using ProxyDroid / iptables with Frida

1. Configure [proxyDroid](https://blog.nviso.eu/2019/08/13/intercepting-traffic-from-android-flutter-applications/) or iptables rules to redirect requests to Burp.

> Execute the command to configure iptables in order to redirect the incoming requests from the application to Burp.

Reviewer 2: ```suggestion
If not using proxyDroid, execute the following commands on the rooted Android device to configure iptables to redirect the incoming requests from the application to Burp:
```

```plaintext
$ iptables -t nat -A OUTPUT -p tcp --dport 80 -j DNAT --to-destination <Your-Proxy-IP>:8080 

$ iptables -t nat -A OUTPUT -p tcp --dport 443 -j DNAT --to-destination <Your-Proxy-IP>:8080 
```

2. Install the [app](../../apps/android/MASTG-APP-0016.md) on the mobile device.

3. Configure the interception proxy. For example, in Burp:

- Under Proxy -> Proxy settings -> Add new Proxy setting.
- Bind listening Port to `8080`.
- Select `Bind to address` to `All interfaces`.
- Request Handling -> support for invisible proxying.

4. Run the [disable-flutter-tls.js](../../tools/generic/MASTG-TOOL-0101.md) frida script.

> ```plaintext

Reviewer 2: Shouldn't these be bash?
